ls1intum/Artemis
# `Programming exercises`: Improve error messages for connection issues in online code editor

### Checklist
#### General


#### Client

### Motivation and Context


Students should not be confused when the connection issues to the Artemis server cause the online editor to malfunction. So the error messages should be improved.

### Description

Whenever an action is not possible in the code editor due to a connection issue an error is displayed at the top right telling the student about this problem.
Additionally when trying to load a file without an internet connection a blank file is shown instead of the last loaded file as before this PR. This blank file is read only.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise with online editor enabled

1. Log in to Artemis
2. Participate using the online editor
3. Loose internet connection by turning of wifi or stopping the server when testing locally
4. Interact with the code editor and check that all actions (submitting, loading an unloaded file, ...) cause an error message at the top right
5. Loading and editing existing files that were loaded before the connection loss should still be possible
6. Reconnect to the server
7. Check that any changes during the connection loss can now be successfully submitted

#### Exam Mode Testing



Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with a Programming Exercise

1. Same as for normal exercises but within an exam

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage




#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [code-editor-ace.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5166/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/ace/code-editor-ace.component.ts.html) | 80.83% | ✅ |
| [code-editor-container.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5166/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/code-editor/container/code-editor-container.component.ts.html) | 85.29% | ✅ |

### Screenshots


![Bildschirmfoto 2023-07-13 um 11 56 36](https://github.com/ls1intum/Artemis/assets/38322605/e3e50cab-eddf-4df6-a5b4-e6fe1dffba6b)
